pkg/pcrbruteforcer/reproduce_event_log.go

```diff
@@ -0,0 +1,195 @@
+			// It matched, everything is OK, let's check next pair.
+			continue
+		}
+
+		// It haven't matched, something is wrong...
+		//
+		// Let's try to remediate if we have an idea how to do that, or
+		// just acknowledge a problem if we cannot remediate it.
+
+		switch {
+		case m.ID == pcr.MeasurementIDPCR0DATA:
+			// If this is the PCR0_DATA measurement then it could be just
+			// a corrupted ACM_POLICY_STATUS register value, let's try
+			// to restore it.
+			correctedACMPolicyStatus, err := bruteForceACMPolicyStatus(*m, imageBytes, ev.Digest.Digest)
+			if err != nil {
+				_ = mErr.Add(fmt.Errorf("PCR0_DATA measurement does not match the digest reported in EventLog and unable to brute force a possible bitflip: %X != %X", mHash[:], ev.Digest.Digest))
+				result = false
+				continue
+			}
+			var buf [8]byte
+			binary.LittleEndian.PutUint64(buf[:], uint64(correctedACMPolicyStatus))
+			_ = mErr.Add(fmt.Errorf("changed ACM_POLICY_STATUS from %X to %X", m.Data[0].ForceData, buf))
+			updatedACMPolicyStatusValue = &correctedACMPolicyStatus
+		default:
+			// I do not know how to remediate this problem.
+			_ = mErr.Add(fmt.Errorf("measurement '%s' does not match the digest reported in EventLog: %X != %X", m.ID, mHash[:], ev.Digest.Digest))
+			result = false
+		}
+	}
+	if evIdx == 0 {
+		// no-one EventLog entry matched, it could not be considered as "remediated"/"fixed".
+		result = false
+	}
+
+	for ; mIdx < len(filteredMeasurements); mIdx++ {
+		m := filteredMeasurements[mIdx]
+		_ = mErr.Add(fmt.Errorf("missing measurement '%s' in EventLog", m.ID))
+	}
+	for ; evIdx < len(events); evIdx++ {
+		ev := events[evIdx]
+		_ = mErr.Add(fmt.Errorf("extra EventLog entry: evIdx == %d: type == %v", evIdx, ev.Type))
+	}
+
+	return result, updatedACMPolicyStatusValue, mErr.ReturnValue()
+}
+
+type bruteForceACMPolicyStatusEventContext struct {
+	Hasher hash.Hash
+	Buf    []byte
+}
+
+func bruteForceACMPolicyStatus(m pcr.Measurement, imageBytes []byte, expectedPCR0DATASHA1Digest []byte) (registers.ACMPolicyStatus, error) {
+	if len(m.Data) == 0 {
+		return 0, fmt.Errorf("no data in the measurement")
+	}
+	if m.ID != pcr.MeasurementIDPCR0DATA {
+		return 0, fmt.Errorf("PCR0_DATA excepted, but received %s", m.ID)
+	}
+	if len(m.Data) == 0 || m.Data[0].ID != pcr.DataChunkIDACMPolicyStatus {
+		return 0, fmt.Errorf("excepted the first data chunk of PCR0_DATA be the ACM_POLICY_STATUS register")
+	}
+
+	m = *m.Copy()
+	acmPolicyStatus := m.Data[0].ForceData
+	if len(acmPolicyStatus) != 8 {
+		return 0, fmt.Errorf("ACM POLICY STATUS register is expected to be 64bits, but it is %dbits", len(acmPolicyStatus)*8)
+	}
+
+	pcr0Data := m.CompileMeasurableData(imageBytes)
+
+	// maxDistance is the maximal hamming distance to brute force to, it is
+	// picked to 6, because it is maximal value for a reasonable brute-force
+	// time (about 1 second).
+	//
+	// For benchmark details see also README.md of package `bruteforcer`.
+	maxDistance := 6
```

Comment:
nit

it looks like it should be configurable (for tools).
+ a good default value should be put into constants